src/coreclr/jit/fgopt.cpp

```diff
@@ -6578,3 +6578,355 @@ void Compiler::fgCompDominatedByExceptionalEntryBlocks()
+        }
+
+        predInfo.Reset();
+
+        // Find the subset of preds that reach along non-critical edges
+        // and populate predInfo.
+        //
+        for (BasicBlock* const predBlock : block->PredBlocks())
+        {
+            if ((predBlock->GetUniqueSucc() == block) && BasicBlock::sameEHRegion(block, predBlock))
+            {
+                Statement* lastStmt = predBlock->lastStmt();
+
+                // Block might be empty.
+                //
+                if (lastStmt == nullptr)
+                {
+                    continue;
+                }
+
+                // Walk back past any GT_NOPs.
+                //
+                Statement* const firstStmt = predBlock->firstStmt();
+                while (lastStmt->GetRootNode()->OperIs(GT_NOP))
+                {
+                    if (lastStmt == firstStmt)
+                    {
+                        // predBlock is evidently all GT_NOP.
+                        //
+                        lastStmt = nullptr;
+                        break;
+                    }
+
+                    lastStmt = lastStmt->GetPrevStmt();
+                }
+
+                // We don't expect to see PHIs but watch for them anyways.
+                //
+                if ((lastStmt != nullptr) && !lastStmt->IsPhiDefnStmt())
+                {
+                    PredInfo info;
+                    info.m_block = predBlock;
+                    info.m_stmt  = lastStmt;
+                    predInfo.Push(info);
+                }
+            }
+        }
+
+        // Are there enough preds to make it interesting?
+        //
+        if (predInfo.Height() < 2)
+        {
+            // Not enough preds to merge
+            return false;
+        }
+
+        // If there are large numbers of viable preds, forgo trying to merge.
+        // While there can be large benefits, there can also be large costs.
+        //
+        // Note we check this rather than countOfInEdges because we don't care
+        // about dups, just the number of unique pred blocks.
+        //
+        if (predInfo.Height() > mergeLimit)
+        {
+            // Too many preds to consider
+            return false;
+        }
+
+        // Find a matching set of preds. Potentially O(N^2) tree comparisons.
+        //
+        int i = 0;
+        while (i < (predInfo.Height() - 1))
+        {
+            matchedPredInfo.Reset();
+            matchedPredInfo.Push(predInfo.TopRef(i));
+            Statement* const baseStmt = predInfo.TopRef(i).m_stmt;
+            for (int j = i + 1; j < predInfo.Height(); j++)
+            {
+                Statement* const otherStmt = predInfo.TopRef(j).m_stmt;
+
+                // Consider: compute and cache hashes to make this faster
+                //
+                if (GenTree::Compare(baseStmt->GetRootNode(), otherStmt->GetRootNode(), true))
+                {
+                    matchedPredInfo.Push(predInfo.TopRef(j));
+                }
```

Comment:
`GenTree::Compare` does not pay attention to `GTF_IND_VOLATILE`, so presumably this can do something like below, which does not seem right.
```
[ p1 ]                  [ p1 ]
[ind<volatile>(addr)]     |
  |                       |
  |  [   p2    ]    -->   |   [ p2 ]
  |  [ind(addr)]          |      |   
  |     |               [ ind(addr) ]
[  block  ]             [   block   ]
```

Reply:
Thanks. 

Was actually expecting to quickly hit more bugs in `GenTree::Compare`, but so far haven't run across any.